common/internal/queues/priority_heap.go

```diff
@@ -0,0 +1,92 @@
+package queues
+
+import (
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+)
+
+// priorityHeap is a priority queue of transactions prioritized by creation time. The oldest transaction is at the front of the queue.
+// It implements the heap interface in the container/heap package.
+type priorityHeap[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+] struct {
+	txs       []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	idToIndex map[int64]int
+}
+
+// newPriorityHeap returns a new priorityHeap instance
+func NewPriorityHeap[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](capacity int) *priorityHeap[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE] {
+	if capacity <= 0 {
+		panic("priority_heap: capacity must be greater than 0")
+	}
+
+	pq := priorityHeap[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		txs:       make([]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], 0, capacity),
+		idToIndex: make(map[int64]int),
+	}
+
+	return &pq
+}
+
+// Close clears the queue
+func (pq *priorityHeap[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Close() {
+	pq.txs = nil
+	pq.idToIndex = nil
+}
+
+// FindIndexByID returns the index of the transaction with the given ID
+func (pq *priorityHeap[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindIndexByID(id int64) (int, bool) {
+	i, ok := pq.idToIndex[id]
+	return i, ok
+}
+
+// Peek returns the next transaction to be processed
+func (pq *priorityHeap[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Peek() *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE] {
+	if len(pq.txs) == 0 {
```

Comment:
/nit is this inconsistent? else where in `priorityHeap` the pattern is to have the caller responsible for providing appropriate inputs based on the state of the queue. Here you sanity check for the caller that the length of the queue is not 0.